### ISSUE TITLE
[Feature] Improve manifest "persist"

### ISSUE THREAD

**Issue by milnak** [ScoopInstaller/Scoop#6179]
#### Is your feature request related to a problem? Please describe.

Manifest "persist" assumes the object to be persisted is a folder, and very often its not.

```json
    "persist": [
        "state.bin",
        "folder"
    ]
```

because of this, almost every manifest forces an empty file to be created, e.g.

```json
    "pre_install": [
        "'state.bin' | ForEach-Object { New-Item -Path (Join-Path -Path $dir -ChildPath $_) -ItemType File -ErrorAction SilentlyContinue | Out-Null }"
    ],

```

#### Describe the solution you'd like

would be great to be able to optionally specify a type, e.g.

```json
    "persist": [
        [ "state.bin", "File" ],
        [ "folder", "Directory" ]
    ]
```

and therefore, scoop wouldn't need to have a pre-existing object in the $dir folder.

**Comment by 29039** [comment]
Honestly I would consider this a bug not an enhancement. The persist function is outright broken when the file to be persisted is a file not a folder.  

The persist function needs to be corrected to be able to specify if the file is supposed to be a file or folder to avoid this "stupid but it works" workaround.

**Comment by BoringBoredom** [comment]
the whole [Nirsoft](https://github.com/ScoopInstaller/Nirsoft) bucket is broken due to this. persist entries ending in file extensions should create files, not folders. this shouldn't be required to handle in manifests

**Comment by BoringBoredom** [comment]
the whole [Nirsoft](https://github.com/ScoopInstaller/Nirsoft) bucket is broken due to this. persist entries ending in file extensions should create files, not folders. it shouldn't be required to handle this in manifests

**Comment by BoringBoredom** [comment]
https://github.com/ScoopInstaller/Scoop/issues/6363

**Comment by HUMORCE** [comment]
> persist entries ending in file extensions should create files, not folders. it shouldn't be required to handle this in manifests

If we just choose between file and dir to created, we're simply spinning our wheels.

Dirs `.configs`, `conf.d` and files `README`, `LICENSE`, `LICENSE_KEY`: 🤷‍♂️

---

Schema that compatible with existing manifests: (DRAFT)

Allows:
- `"persist": "app.conf"` | `"persist": ["config.ini", "config"]` - Current behavior
- `"persist": { "file": "app.conf", "directory": ["conf", "conf.d"] }`

```
"definitions": {
...
"persist": {
    "anyOf": [
        {
            "$ref": "#/definitions/stringOrArrayOfStringsOrAnArrayOfArrayOfStrings"
        },
        {
            "additionalProperties": false,
            "properties": {
                "file": {
                    "$ref": "#/definitions/stringOrArrayOfStrings"
                },
                "directory": {
                    "$ref": "#/definitions/stringOrArrayOfStrings"
                }
            },
            "anyOf": [
                {
                    "required": [
                        "file"
                    ]
                },
                {
                    "required": [
                        "directory"
                    ]
                }
            ],
            "type": "object"
        }
    ]
},
...
"persist": {
    "$ref": "#/definitions/persist"
},
```

**Comment by BoringBoredom** [comment]
> If we just choose between file and dir to created, we're simply spinning our wheels.
> 
> Dirs `.configs`, `conf.d` and files `README`, `LICENSE`, `LICENSE_KEY`: 🤷‍♂️

imo, it's far easier to handle some edge cases than spam file creation code in hundreds, if not thousands, of manifests

**Comment by BoringBoredom** [comment]
> If we just choose between file and dir to created, we're simply spinning our wheels.
> 
> Dirs `.configs`, `conf.d` and files `README`, `LICENSE`, `LICENSE_KEY`: 🤷‍♂️

imo, it's far easier to handle some edge cases than spam file creation code in hundreds, if not thousands, of manifests. defaulting to inferring type from extensions, but also allowing explicit definition of `dir` or `file`, sounds sensible, too

**Comment by 29039** [comment]
> > If we just choose between file and dir to created, we're simply spinning our wheels.
> > Dirs `.configs`, `conf.d` and files `README`, `LICENSE`, `LICENSE_KEY`: 🤷‍♂️
> 
> imo, it's far easier to handle some edge cases than spam file creation code in hundreds, if not thousands, of manifests. defaulting to inferring type from extensions, but also allowing explicit definition of `dir` or `file`, sounds sensible, too

I did a search with Everything app for *.json %userprofile%\scoop\buckets\ content:persist and across buckets of extras, java, main, versions, nirsoft, I got 1303 results. I then fed those files into a powershell script I made in deepseek to get a bit of an analysis of the persist contents:

=== DETAILED PERSIST ANALYSIS ===
Files with persist entries: 1303
Total persist entries found: 2497

Entry Type Breakdown:
- No dots (likely folders): 1165 (46.7%)
- Short extensions (1-2 chars): 13 (0.5%)
- Standard extensions (3 chars): 1059 (42.4%)
- Long extensions (4+ chars): 240 (9.6%)
- Complex structures: 20 (0.8%)

(complex structures meaning extra nesting was added)

Perhaps the most straightforward solution would be to either change "persist" to assume a directory or keep it as-is (which already assumes a directory anyway), or assume a dir only if it doesn't end with a dot followed by 3 letters (otherwise it's probably a file)

Then we add new two separate types: "persistfiles" and "persistdirs" to the schema which is more explicit.

and then we can focus on 
1. whats reported broken (nirsoft wakemeonlan for me) 
2. transitioning packages with a complex persist (unnecessary extra nesting) = 20
3. transitioning packages with file extensions less than 3 letters = 13
3. transitioning packages with file extensions more than 3 letters = 240

273 manual manifest changes seems quite manageable, especially with knowing while files have the required changes.

Then we end up under the old scheme with 1165 probable directories (no dots in the filename) and 1059 probable files (dot followed by 3 characters in the file name.

I also did an extension analysis and got:

top extensions:
- .ini: 548 (51.7%)
- .cfg: 319 (30.1%)
- .xml: 71 (6.7%)
- .txt: 44 (4.2%)
- .dat: 20 (1.9%)
- .log: 10 (0.9%)
- .key: 4 (0.4%)
- .yml: 4 (0.4%)
- .ps1: 4 (0.4%)
- .lua: 4 (0.4%)
- .pem: 3 (0.3%)

These files were hit twice, over 10 files: .lpl .pls .bar .br2 .bin
These files were uniquely hit over 18 files: .cnf .p2b .nsh .db3 .dfu .pth .sdf .inf .pid .lss .dll .zip .qdr .csv .vnc .bak .vst .nss 

From the top 6, 5 of them are almost definitely a (plain text) file. a .dat probably is a file too, but maybe not so more risk given the meaning of dat could be "user data", so I'm not including that in assumptions

So if we want to be even more safe about assuming that a persist is referring to a file, we could count only ending in .ini .cfg .xml .txt .log 

Then that would give us 10+18+3+4+4+4+4+20=67 more manual fixes to ensure that it is properly captured by persist files as uncommon+dat file types.

total manual checks/fixes for edge cases: 340
safe directory assumptions (no dots): 1165
safe file assumptions (ending in .ini .cfg .xml .txt or .log): 1059
likely that there will be some persist "files" which are actually extensionless. so that is going to be pretty much undetectable, so we'd have to monitor for bug reports or gradually review the actual functioning of those programs to verify. I'd lean towards waiting for bug reports because likely not actually breaking anything if not being reported (although unreported bugs because the user doesn't know enough to report them are annoying).

[deepseek_powershell_20250521_f4446e.ps1.txt](https://github.com/user-attachments/files/20360001/deepseek_powershell_20250521_f4446e.ps1.txt)

**Comment by HUMORCE** [comment]
The point is compatibility, we lack active maintainers/contributor, whether Scoop Core or Buckets.

We should make it as perfect as possible by one-time instead of causing known issues that cannot located. 3rd-party/personal/private buckets should be consider at same time.

The possible solutions posted above as example, contribution welcome.

**Comment by 29039** [comment]
> The point is compatibility, we lack active maintainers/contributor, whether Scoop Core or Buckets.
> 
> We should make it as perfect as possible by one-time instead of causing known issues that cannot located.

I would suggest that the most compatible solution is to leave "persist" as-is which will not add any further bugs to what we already have, but we should also add "persistfiles" and "persistdirs" parameters to be able to be more specific without breaking compatibility. 

To my knowledge scoop automatically updates itself and would therefore understand the new parameters.

We could then work on a task to update particular manifests to make use of the new parameters to reduce the chances of bugs, while also being able to detect which has been fixed to the new system or not by no longer having "persist" in the manifest if it is updated.

If you are willing to add a small amount of risk in potential new bugs, which I think is quite small, but still an increased risk, in addition to the the new parameters, we could also slightly modify the existing "persist" behaviour to auto detect .ini .cfg .xml .txt .log as being files. But this is not a necessary risk. 

I would support adding new parameters of "persistdirs" and "persistfiles" in addition to the old as being the minimum viable solution.

There are other suggestions on changing the behaviour and schema of the existing "persist" but I argue against this, it is my risky, more work to fix old manifest, more complex code and structure, having to repeat if it is a file or dir constantly for each item. I think that separate persistdirs and persistfiles is a clean separation from the old code.

**Comment by BoringBoredom** [comment]
can you point me to a manifest that this PR would break? https://github.com/ScoopInstaller/Scoop/pull/6364/files

the way i see it, there can be a few situations:
- existing installations are completely unaffected as the dirs/files are already existent (**_same behavior as right now_**)
- new installations where persist doesn't contain file extensions will create dirs (**_same behavior as right now_**)
- new installations where persist contains files that don't have an extension will create dirs (**_same behavior as right now_**)
- new installations where persist contains file extensions will create empty files (**_fixes hundreds of broken manifests immediately_**)
- new installations where persist contains folders that are named like files will create files (**_this would break, i'd guesstimate, very few manifests_**)

this fixes countless broken manifests immediately, alleviating the problem until a proper spec is drafted and implemented. this is a good trade-off, in my opinion.

for example, as of now it looks like the whole Nirsoft bucket is broken, as none of the manifests actually handle file creation: https://github.com/search?q=repo%3AScoopInstaller%2FNirsoft%20pre_install&type=code

**Comment by 29039** [comment]
> can you point me to a manifest that this PR would break? https://github.com/ScoopInstaller/Scoop/pull/6364/files

I haven't looked too thoroughly through thousands of files but this one could break because tunnels.d is supposed to be a directory:
https://github.com/ScoopInstaller/Extras/blob/master/bucket/i2pd.json

> * existing installations are completely unaffected as the dirs/files are already existent (**_same behavior as right now_**)
updates would run the manifest again?

> * new installations where persist doesn't contain file extensions will create dirs (**_same behavior as right now_**)
Yes

> * new installations where persist contains files that don't have an extension will create dirs (**_same behavior as right now_**)
You already said that

> * new installations where persist contains file extensions will create empty files (**_fixes hundreds of broken manifests immediately_**)
But potentially breaks hundreds of broken manifests which were only tested with broken manifests

> * new installations where persist contains folders that are named like files will create files (**_this would break, i'd guesstimate, very few manifests_**)
perhaps few, but I already found an example, it would be a really tough job going through all the manifests and guessing, testing, looking up it's documentation to figure out if it is a dir or file. I would be a whole lot more confident if it is a fixed set of extensions which are pretty much guaranteed to be files like .txt .ini .json .cfg etc. It would be incredibly unusual to see a directory with a common file extension. This would at least workaround the nirsoft in the short term - which all use .cfg and .ini

> this fixes countless broken manifests immediately, alleviating the problem until a proper spec is drafted and implemented. this is a good trade-off, in my opinion.
Bad tradeoff, because it doesn't fundamentally fix the limitation, possibly even putting off a proper implementation. Should fix it properly in the first place by updating the spec with "persistdirs" and "persistfiles" which can pretty much re-use the existing persist code with changes for not much extra work, if adding to the spec is permitted. A PR which will do the entire fix, with the new spec, would be so much better and just fix everything at once.

> for example, as of now it looks like the whole Nirsoft bucket is broken, as none of the manifests actually handle file creation: 
I could reluctantly accept your PR if it was limited to only .cfg & .ini, to at least get nirsoft tangibly working properly in the short term,  but there would need to be some kind of assurance that a proper fix to the spec would be coming as well. 

But even that comes with risk, .cfg could affect teraterm, vortexdm, wavosaur (could be more)

The other alternative is to just fix nirsoft manifests to create the files as part of "pre_install" like other manifests are doing, until there is a new spec.

**Comment by BoringBoredom** [comment]
> I haven't looked too thoroughly through thousands of files but this one could break because tunnels.d is supposed to be a directory:
> 
> https://github.com/ScoopInstaller/Extras/blob/master/bucket/i2pd.json

good catch. i suppose extension detection should be narrowed down to 2-4 chars

> > * existing installations are completely unaffected as the dirs/files are already existent (**_same behavior as right now_**)
> 
> updates would run the manifest again?

they already do in the current form from what i can tell. the updater in `scoop-update.ps1` calls the function `install_app` from `install.ps1` which calls the function `persist_data` (the function the PR targets)

> > * new installations where persist contains files that don't have an extension will create dirs (**_same behavior as right now_**)
> 
> You already said that

this is a different scenario. i should've been more explicit. the former scenario is directory without extension, the latter scenario is file without extension

> > * new installations where persist contains file extensions will create empty files (**_fixes hundreds of broken manifests immediately_**)
> 
> But potentially breaks hundreds of broken manifests which were only tested with broken manifests, not to mention all the custom pre_install code which could conflict. safest option is to leave the old persist code as-is and migrate manifests to the new spec over time

the existing `pre_install` code creating folders and files are unaffected as the code path in the PR never triggers if dirs/files are already present

> > * new installations where persist contains folders that are named like files will create files (**_this would break, i'd guesstimate, very few manifests_**)
> 
> perhaps few, but I already found an example, it would be a really tough job going through all the manifests and guessing, testing, looking up it's documentation to figure out if it is a dir or file. I would be a whole lot more confident if it is a fixed set of extensions which are pretty much guaranteed to be files like .txt .ini .json .cfg etc. It would be incredibly unusual to see a directory with a common file extension. This would at least workaround the nirsoft in the short term - which all use .cfg and .ini

i think the trade-off is still worth it considering how many manifests are currently broken due to this

> > this fixes countless broken manifests immediately, alleviating the problem until a proper spec is drafted and implemented. this is a good trade-off, in my opinion.
> 
> Bad tradeoff, because it doesn't fundamentally fix the limitation, possibly even putting off a proper implementation. Should fix it properly in the first place by updating the spec with "persistdirs" and "persistfiles" which can pretty much re-use the existing persist code with changes for not much extra work, if adding to the spec is permitted. A PR which will do the entire fix, with the new spec, would be so much better and just fix everything at once.

that won't fix the currently broken manifests. who is going to fix all those?

> > for example, as of now it looks like the whole Nirsoft bucket is broken, as none of the manifests actually handle file creation:
> 
> I could reluctantly accept your PR if it was limited to only .cfg & .ini, to at least get nirsoft tangibly working properly in the short term, but there would need to be some kind of assurance that a proper fix to the spec would be coming as well.
> 
> But even that comes with risk, .cfg could affect teraterm, vortexdm, wavosaur (could be more)
> 
> The other alternative is to just fix nirsoft manifests to create the files as part of "pre_install" like other manifests are doing, until there is a new spec.

i agree this isn't a real fix but considering just how many manifests are currently broken i think it's reasonable to implement a temporary workaround that heavily reduces the amount of total currently broken manifests until a spec is implemented

**Comment by 29039** [comment]
@BoringBoredom I can offer a compromise solution which does not add many lines of code compared to reimplementing persist very heavily.

In your PR just make the following changes:

Line 1026:
```
            if ($source -match '^file:') {
                $isFile = $true
                $source = $source -replace '^file:', ''
                $target = $target -replace '^file:', ''
            }
```

Line 1041:
`                if ($isFile -eq $true) {`

This way it is very simple, to specify that a file is a file, just prefix it with **file:** (lower case)

Because it contains a colon, it would be an invalid filename to collide with any real files, nor is a four letter word able to be a drive letter. 

Then we can go into all the nirsoft manifests and prefix **file:** to each one.

Using an AI like DeepSeek to make a powershell script, this is very easy to update in bulk

As I posted a powershell script a few comments ago, it is actually super simple to make a script to do an analysis of a lot of manifest files at once to figure out where the likely persist files are. That was how I could find examples so quickly.

https://github.com/ScoopInstaller/Scoop/issues/6179#issuecomment-2896513535

It's a big but not impossible task to find all the likely file candidates, A lot are already handling it in the pre_install manually, in which case it's not even broken but we could simplify. others we flag for review as possibly broken if it is not being handled by pre_install but looks like a file

can even update in bulk using an AI made powershell script as well.

**Comment by 29039** [comment]
No, I am looking through and the vast majority of manifests using persist on probable files are using pre_install to create the files first. Nirsoft is the outlier. There is a large amount of testing needed to work out the exact interaction of the detection code and all the custom pre_install scripts, and I don't think that the juice is worth the squeeze for the amount of risk being introduced. At this point I'd rather just update all the nirsoft with pre_install than to continue down this path of a guessing script of unknown interaction with all the different packages and pre_install. For example, a pre_install might wipe the persisted files because it's unaware of this new detection. There is a lot of breakage possible with double handling. 

The absolute best thing possible is to have a proper implementation to handle files explicitly and use AI tools & scripts to review & update manifests in bulk .

**Comment by HUMORCE** [comment]
I made a PR for the possible solutions I left earlier. it will not fixes any manifests. all behavior consistent with current. persistence error should fixes on manifest instead of core.

**Comment by anacude** [comment]
> The point is compatibility, we lack active maintainers/contributor, whether Scoop Core or Buckets.
> 
> We should make it as perfect as possible by one-time instead of causing known issues that cannot located. 3rd-party/personal/private buckets should be consider at same time.
> 
> The possible solutions posted above as example, contribution welcome.

I'm a freelance developer looking to contribute. What are the current high-priority tasks I can help with? @HUMORCE
